core/src/com/unciv/logic/map/MapGenerator.kt

```diff
@@ -96,6 +101,49 @@ class MapGenerator(val ruleset: Ruleset) {
         }
     }
 
+    /**
+     * generates a single river at the center of the map.
+     * this is a proof of concept to show that the spawning of rivers works
+     * and other aspects of the game can be adjusted to fully incorporate them.
+     */
+    private fun createRivers(map: TileMap) {
+        map.rivers = ArrayList(Collections.singletonList(createRiver(Vector2(0F, 0F), map)))
+    }
+
+    private fun createRiver(source: Vector2, map: TileMap): River {
+        val river = River()
+        var currentRiverTiles = generateFirstRiverPair(source)
+        var length = 0
+        while (areLegalRiverTiles(currentRiverTiles, map) && length < MAX_RIVER_LENGTH) {
+            length++
+            river.course.add(currentRiverTiles)
+            currentRiverTiles = generateNextRiverPair(currentRiverTiles)
+        }
+        return river
+    }
+
+    private fun generateFirstRiverPair(source: Vector2): Pair<Vector2, Vector2> {
+        return Pair(source, Vector2(source).add(1F, 0F))
+    }
+
+    private fun generateNextRiverPair(oldPair: Pair<Vector2, Vector2>): Pair<Vector2, Vector2> {
+        val newSecond: Vector2 = oldPair.first
+        val newFirst = if (oldPair.first.x + 1 == oldPair.second.x) {
+            Vector2(newSecond).sub(0F, 1F)
+        } else {
+            Vector2(newSecond).sub(1F, 0F)
+        }
+        return Pair(newFirst, newSecond)
+    }
+
+    private fun areLegalRiverTiles(currentRiverTiles: Pair<Vector2, Vector2>, map: TileMap): Boolean {
+        return (!(map[currentRiverTiles.first].isWater
+                || map[currentRiverTiles.second].isWater)
+                && map.contains(currentRiverTiles.first)
+                && map.contains(currentRiverTiles.second))
```

Comment:
I might be missing something, but shouldn't you check the contains before you use it as indices for get()?